modelling/src/main/java/org/axonframework/modelling/command/StatefulCommandHandlingComponent.java

```diff
@@ -0,0 +1,308 @@
+import org.axonframework.commandhandling.CommandMessage;
+import org.axonframework.commandhandling.CommandResultMessage;
+import org.axonframework.commandhandling.SimpleCommandHandlingComponent;
+import org.axonframework.common.infra.ComponentDescriptor;
+import org.axonframework.common.infra.DescribableComponent;
+import org.axonframework.messaging.MessageStream;
+import org.axonframework.messaging.QualifiedName;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link CommandHandlingComponent} implementation which allows for stateful handling of commands.
+ * <p>
+ * Models can be registered to this component through the {@link #registerModel(String, Class, Function, ModelLoader)}
+ * method. These models can be resolved during the handling of commands by the {@link StatefulCommandHandler StatefulCommandHandlers}
+ * that are subscribed to this component. The models are resolved based on the command that is being handled.
+ * <p>
+ * Can load the models eagerly, meaning that all models are loaded when the command is handled and the resolved
+ * identifier is non-null, or lazily, meaning that
+ * the models are only loaded when they are requested.
+ *
+ * @author Mitchell Herrijgers
+ * @since 5.0.0
+ */
+public class StatefulCommandHandlingComponent implements
+        CommandHandlingComponent,
+        StatefulCommandHandlerRegistry<StatefulCommandHandlingComponent>,
+        DescribableComponent {
+
+    private final String name;
+    private final SimpleCommandHandlingComponent handlingComponent;
+    private final List<ModelDefinition<?, ?>> modelDefinitions = new CopyOnWriteArrayList<>();
+    private boolean loadModelsEagerly = false;
+
+    /**
+     * Initializes a new stateful command handling component with the given {@code name}.
+     * @param name The name of the component
+     */
+    private StatefulCommandHandlingComponent(String name) {
+        this.name = name;
+        this.handlingComponent = SimpleCommandHandlingComponent.forComponent(name);
+    }
+
+    /**
+     * Creates a new stateful command handling component with the given {@code name}.
+     * @param name The name of the component
+     * @return A new stateful command handling component with the given {@code name}
+     */
+    public static StatefulCommandHandlingComponent forName(String name) {
+        return new StatefulCommandHandlingComponent(name);
+    }
+
+    /**
+     * Tells this component to load models eagerly. This means that all models are loaded when the command is handled
+     * and the resolved identifier is non-null.
+     *
+     * @return this component for fluent interfacing
+     */
+    public StatefulCommandHandlingComponent loadModelsEagerly() {
```

Comment:
I think it could be an enum in the constructor parameter (you can add two overloads, and set the default). Such fluent interfaces can be placed in configuration part.
Why? For me it's weir that I can change the modes while component is "active", I mean even after creation.